federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -724,18 +724,24 @@ func (s *ServiceController) lockedUpdateDNSRecords(service *cachedService, clust
 	if !wantsDNSRecords(service.appliedState) {
 		return nil
 	}
-	ensuredCount := 0
+
+	var err error
+	missedCount := 0
+	unensuredCount := 0
 	for key := range s.clusterCache.clientMap {
 		for _, clusterName := range clusterNames {
 			if key == clusterName {
-				s.ensureDnsRecords(clusterName, service)
-				ensuredCount += 1
+				if err = s.ensureDnsRecords(clusterName, service); err != nil {
+					unensuredCount += 1
+				}
+			} else {
+				missedCount += 1
 			}
 		}
 	}
-	if ensuredCount < len(clusterNames) {
-		return fmt.Errorf("Failed to update DNS records for %d of %d clusters for service %v due to missing clients for those clusters",
-			len(clusterNames)-ensuredCount, len(clusterNames), service)
+	if missedCount > 0 || unensuredCount > 0 {
+		return fmt.Errorf("Failed to update DNS records for %d clusters for service %v due to missing clients[missed count: %d] or ensuring DNS records error[unensured count: %d] %v",
+			len(clusterNames), service, missedCount, unensuredCount, err)
```

Comment:
According to comments in last commit, code is changed to this.  Here I don't return the error of missing clients or unensured DNS records separately, because they may occur simultaneously, i.e. some clients may miss, and the not missed clients may fail to ensure DNS records.

/cc @nikhiljindal @quinton-hoole @jfrazelle 

by nikhiljindal 
>The error 2 lines below is not correct now. We need to keep track of number of missing cluster clients and number of failed ensureDNsRecords and return an error accordingly.

by xiangpengzhao
> Reasonable. The numbers may be counted by below?
1. number of missing cluster clients: count of key != clusterName
2. number of failed ensureDNsRecords: count of when s.ensureDnsRecords return non nil?
If so, we can define two new variables: missedCount, unensuredCount(instead current ensuredCount)